### ISSUE TITLE
Bug: Conditional reshaping prevents proper empty tensor shape transformation.

### ISSUE THREAD

**Issue by WillieMaddox** [ultralytics/ultralytics#21119]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Predict

## Description
The current conditional logic for reshaping keypoint predictions fails to properly reshape empty tensors, resulting in incorrect output dimensions.  In `models/yolo/pose/predict.py`, if the `pred` argument in `construct_result` is empty:

## Current Behavior
```python
# pred: torch.tensor([], device='cuda:0', size=(0, 14))  # empty
pred_kpts = pred[:, 6:].view(len(pred), *self.model.kpt_shape) if len(pred) else pred[:, 6:]
# pred_kpts: tensor([], device='cuda:0', size=(0, 8))  # WRONG
pred_kpts = ops.scale_coords(img.shape[2:], pred_kpts, orig_img.shape)
result.update(keypoints=pred_kpts)
# result.keypoints.data: tensor([], device='cuda:0', size=(1, 0, 8))  # WRONG
```

## Expected Behavior
```python
# pred: torch.tensor([], device='cuda:0', size=(0, 14))  # empty
pred_kpts = pred[:, 6:].view(len(pred), *self.model.kpt_shape)
# pred_kpts: tensor([], device='cuda:0', size=(0, 4, 2))  # CORRECT (assuming kpt_shape=[4, 2])
pred_kpts = ops.scale_coords(img.shape[2:], pred_kpts, orig_img.shape)
result.update(keypoints=pred_kpts)
# result.keypoints.data: tensor([], device='cuda:0', size=(0, 4, 2))  # CORRECT
```

## Root Cause
The conditional check `if len(pred)` evaluates to `False` for empty tensors (len=0), causing the code to execute the `else` branch which only slices (`pred[:, 6:]`) without reshaping.

### Environment

$ yolo checks
Ultralytics 8.3.155 🚀 Python-3.10.17 torch-2.7.1+cu126 CUDA:0
Setup complete ✅

|Setup| |
|-|-|
|OS                  |Linux-6.11.0-26-generic-x86_64-with-glibc2.39
|Environment         |Linux
|Python              |3.10.17
|Install             |pip
|CUDA                |12.6

|Package |Version |
|----------------|--------------|
|numpy               |✅ 2.2.6>=1.23.0|
|matplotlib          |✅ 3.10.3>=3.3.0|
|opencv-python       |✅ 4.11.0.86>=4.6.0|
|pillow              |✅ 10.4.0>=7.1.2|
|pyyaml              |✅ 6.0.2>=5.3.1|
|requests            |✅ 2.32.4>=2.23.0|
|scipy               |✅ 1.15.3>=1.4.1|
|torch               |✅ 2.7.1>=1.8.0|
|torch               |✅ 2.7.1!=2.4.0,>=1.8.0; sys_platform == "win32"|
|torchvision         |✅ 0.22.1>=0.9.0|
|tqdm                |✅ 4.67.1>=4.64.0|
|psutil              |✅ 7.0.0|
|py-cpuinfo          |✅ 9.0.0|
|pandas              |✅ 2.3.0>=1.1.4|
|ultralytics-thop    |✅ 2.0.14>=2.0.0|

### Minimal Reproducible Example

1. Create an empty tensor: `pred = torch.tensor([]).reshape(0, 14)`
2. Set `kpt_shape = [4, 2]`
3. Run the current code: `pred_kpts = pred[:, 6:].view(len(pred), *kpt_shape) if len(pred) else pred[:, 6:]`
4. Check shape: `pred_kpts.shape` returns `torch.Size([0, 8])` instead of `torch.Size([0, 4, 2])`

## Proposed Fix
Remove the unnecessary conditional check since PyTorch handles empty tensor reshaping correctly:

```python
# Current (buggy)
pred_kpts = pred[:, 6:].view(len(pred), *self.model.kpt_shape) if len(pred) else pred[:, 6:]

# Fixed
pred_kpts = pred[:, 6:].view(len(pred), *self.model.kpt_shape)
```

## Alternative Fixes
If the conditional is needed for other reasons:

```python
# Option 1: Fix the conditional logic
pred_kpts = pred[:, 6:].view(len(pred), *kpt_shape) if len(pred) > 0 else pred[:, 6:].view(0, *kpt_shape)

# Option 2: Use explicit empty tensor creation
pred_kpts = pred[:, 6:].view(-1, *kpt_shape) if pred.numel() > 0 else torch.empty(0, *kpt_shape, device=pred.device, dtype=pred.dtype)
```

## Impact
- Downstream code expecting `(0, K, 2)` shaped tensors will receive `(1, 0, K*2)` shaped tensors in `result.keypoints.data`.
- This can cause dimension mismatches in subsequent operations
- Affects any batch processing pipeline that may encounter empty batches

## Test Case
```python
import torch

def test_empty_tensor_reshape_4_2():
    pred = torch.tensor([]).reshape(0, 14)
    kpt_shape = [4, 2]
    
    # Current buggy behavior
    pred_kpts_wrong = pred[:, 6:].view(len(pred), *kpt_shape) if len(pred) else pred[:, 6:]
    assert pred_kpts_wrong.shape == torch.Size([0, 8])  # Current wrong result
    
    # Expected correct behavior
    pred_kpts_right = pred[:, 6:].view(len(pred), *kpt_shape)
    assert pred_kpts_right.shape == torch.Size([0, 4, 2])  # Expected result
    
    # Verify it still works for non-empty tensors
    pred_filled = torch.randn(3, 14)
    pred_kpts_filled = pred_filled[:, 6:].view(len(pred_filled), *kpt_shape)
    assert pred_kpts_filled.shape == torch.Size([3, 4, 2])
    
    print("All tests passed!")

test_empty_tensor_reshape_4_2()
```

The same is true for COCO human pose style keypoints:

```python
def test_empty_tensor_reshape_17_3():
    pred = torch.tensor([]).reshape(0, 57)
    kpt_shape = [17, 3]

    # Current buggy behavior
    pred_kpts_wrong = pred[:, 6:].view(len(pred), *kpt_shape) if len(pred) else pred[:, 6:]
    assert pred_kpts_wrong.shape == torch.Size([0, 51])  # Current wrong result

    # Expected correct behavior
    pred_kpts_right = pred[:, 6:].view(len(pred), *kpt_shape)
    assert pred_kpts_right.shape == torch.Size([0, 17, 3])  # Expected result

    # Verify it still works for non-empty tensors
    pred_filled = torch.randn(3, 57)
    pred_kpts_filled = pred_filled[:, 6:].view(len(pred_filled), *kpt_shape)
    assert pred_kpts_filled.shape == torch.Size([3, 17, 3])

    print("All tests passed!")

test_empty_tensor_reshape_17_3()
```

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @WillieMaddox, thank you for reporting this and for providing such a detailed description and a minimal reproducible example (MRE) 🚀! This is an automated response to help guide your issue, and an Ultralytics engineer will also assist you soon.

We recommend new users visit the [Docs](https://docs.ultralytics.com/) for general guidance, where you'll find [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples. For bug reports like this one, thanks for including an MRE—this is very helpful for debugging 🐛.

If you're interested in community support or real-time discussions, you can join us on [Discord](https://discord.com/invite/ultralytics) 🎧, participate in [Discourse](https://community.ultralytics.com/), or share insights on our [Subreddit](https://reddit.com/r/Ultralytics).

## Upgrade

Before proceeding, please make sure you are using the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of these up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you again for your thorough report and willingness to submit a PR! An Ultralytics engineer will review your suggestions and follow up soon. 😊

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in the linked PR #21120. This update standardizes keypoint prediction handling within YOLO pose models by removing unnecessary conditional checks and consistently using `.view()` for tensor reshaping. These improvements ensure more reliable and predictable keypoint extraction across both prediction and validation, which should help eliminate shape-related bugs and enhance the overall stability of pose estimation tasks.

Special thanks to @WillieMaddox for authoring the PR and to @Laughing-q for their valuable contributions.

To test if this resolves the issue, try installing the latest version directly from the repository:

```
pip install git+https://github.com/ultralytics/ultralytics.git@main
```

Alternatively, you can wait for the next official release on PyPI, which will include these changes.

Feedback is very welcome—please let us know if the update resolves your issue or if you notice any other related problems. Thank you for reporting this and helping improve the Ultralytics ecosystem! 🙏
